rust/psibase_macros/psibase-macros-lib/src/service_macro/actions.rs

```diff
@@ -190,3 +192,108 @@ pub fn process_action_schema(
         actions.insert(#name_str.to_string(), #psibase_mod::fracpack::FunctionType{ params: builder.insert::<#action_mod::#name>(), result: #ret });
     }
 }
+
+pub struct PreAction {
+    pub fn_name: Option<Ident>,
+    pub exclude: PathList,
+}
+
+impl PreAction {
+    pub fn has_pre_action(&self) -> bool {
+        self.fn_name.is_some()
+    }
+}
+
+impl Default for PreAction {
+    fn default() -> Self {
+        PreAction {
+            fn_name: None,
+            exclude: PathList::from(vec![]),
+        }
+    }
+}
+
+fn is_pre_action_attr(attr: &Attribute) -> bool {
+    AttrStyle::Outer == attr.style && attr.meta.path().is_ident("pre_action")
+}
+
+#[derive(Debug, FromMeta)]
+struct PreActionOptions {
+    exclude: PathList,
+}
+
+pub fn check_for_pre_action(pre_action_info: &mut PreAction, items: &mut Vec<Item>) {
+    let mut num_pre_action_fns = 0;
+    for item in items.iter_mut() {
+        if let Item::Fn(f) = item {
+            let pre_action_attrs = f
+                .attrs
+                .clone()
+                .into_iter()
+                .filter(|attr| is_pre_action_attr(attr))
+                .collect::<Vec<Attribute>>();
+            let num_pa_attrs_on_fn = pre_action_attrs.len();
+            if num_pa_attrs_on_fn == 0 {
+                continue;
+            }
+            if num_pa_attrs_on_fn > 1 {
+                abort!(item, "More than one pre_action attribute found.");
```

Comment:
For all of these errors, it would be better to insert a `compile_error!` with an appropriate span (or return `syn::Error`) so the message shows the actual location.

Reply:
@swatanabe Ok, I think I'm happy with this implementation. Let me know if you have further feedback. I've also tag you on another piece of this I feel should be nicer but can't find a way to simplify.